src/SignalR/common/Shared/ClientResultsManager.cs

```diff
@@ -0,0 +1,195 @@
+
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using Microsoft.AspNetCore.SignalR.Protocol;
+
+namespace Microsoft.AspNetCore.SignalR.Internal;
+
+// Common type used by our HubLifetimeManager implementations to manage client results.
+// Handles cancellation, cleanup, and completion, so any bugs or improvements can be made in a single place
+internal class ClientResultsManager : IInvocationBinder
+{
+    private readonly ConcurrentDictionary<string, (Type Type, string ConnectionId, object Tcs, Func<object, CompletionMessage, Task> Completion)> _pendingInvocations = new();
+
+    public Task<T> AddInvocation<T>(string connectionId, string invocationId, CancellationToken cancellationToken)
+    {
+        var tcs = new TaskCompletionSourceWithCancellation<T>(this, connectionId, invocationId, cancellationToken);
+        var result = _pendingInvocations.TryAdd(invocationId, (typeof(T), connectionId, tcs, static (state, completionMessage) =>
+        {
+            var tcs = (TaskCompletionSourceWithCancellation<T>)state;
+            if (completionMessage.HasResult)
+            {
+                tcs.SetResult((T)completionMessage.Result);
+            }
+            else
+            {
+                tcs.SetException(new Exception(completionMessage.Error));
+            }
+            return Task.CompletedTask;
+        }
+        ));
+        Debug.Assert(result);
+
+        tcs.RegisterCancellation();
+
+        return tcs.Task;
+    }
+
+    public void AddInvocation(string invocationId, (Type Type, string ConnectionId, object Tcs, Func<object, CompletionMessage, Task> Completion) invocationInfo)
+    {
+        var result = _pendingInvocations.TryAdd(invocationId, invocationInfo);
+        Debug.Assert(result);
+    }
+
+    public Task TryCompleteResult(string connectionId, CompletionMessage message)
+    {
+        if (_pendingInvocations.TryGetValue(message.InvocationId!, out var item))
+        {
+            if (item.ConnectionId != connectionId)
+            {
+                throw new InvalidOperationException($"Connection ID '{connectionId}' is not valid for invocation ID '{message.InvocationId}'.");
+            }
+
+            // if false the connection disconnected right after the above TryGetValue
+            // or someone else completed the invocation (likely a bad client)
+            // we'll ignore both cases
+            if (_pendingInvocations.Remove(message.InvocationId!, out _))
+            {
+                return item.Completion(item.Tcs, message);
+            }
+        }
+        else
+        {
+            // connection was disconnected or someone else completed the invocation
+        }
+        return Task.CompletedTask;
+    }
+
+    public (Type Type, string ConnectionId, object Tcs, Func<object, CompletionMessage, Task> Completion)? RemoveInvocation(string invocationId)
```

Comment:
```suggestion
    public (Type Type, string ConnectionId, object Tcs, Func<object, CompletionMessage, Task> CompleteAsync)? RemoveInvocation(string invocationId)
```

Unless we turn it into an Action, then `Complete`.